Assets/Scripts/R&D Tree/ResearchNode.cs

```diff
@@ -0,0 +1,134 @@
+
+    private float _timer;
+
+    public UnityEvent ResearchDoneEvent = new UnityEvent();
+
+    private ResearchNodeSetting _nodeSetting;
+
+    public SkillState CurrentSkillState;
+
+    private void OnEnable()
+    {
+        SetUIForStates();
+    }
+
+    private void Start()
+    {
+        _nodeSetting = GetComponent<ResearchNodeSetting>();
+
+        _researchBar.maxValue = _nodeSetting.ResearchTime;
+        AddListenersToEvent();
+    }
+
+    /// <summary>
+    /// This function is called when the researchnode started to be researched
+    /// </summary>
+    public void TimerForInDevelopment()
+    {
+        if (CurrentSkillState != SkillState.inDevelopment) return;
+
+        ///TODO: Check if there is enough money
+
+        _timer += Time.fixedDeltaTime;
+        _researchBar.value = _timer;
+        if (_timer >= _nodeSetting.ResearchTime)
+        {
+            ResearchDoneEvent.Invoke();
+        }
+    }
+
+    /// <summary>
+    /// This function is called when the research is done
+    /// </summary>
+    private void ResearchDone()
+    {
+        CurrentSkillState = SkillState.bought;
+
+        if (RDTreeManager.Instance.CurrentResearching != null)
+        {
+            RDTreeManager.Instance.CurrentResearching = null;
+        }
+    }
+
+    /// <summary>
+    /// This function is called when the start research button is clicked
+    /// </summary>
+    public void OnClick()
+    {
+        if (CurrentSkillState == SkillState.available)
+        {
+            CurrentSkillState = SkillState.inDevelopment;
+            RDTreeManager.Instance.ChooseNewResearch(this);
+            SetUIForStates();
+        }
+    }
+
+    /// <summary>
+    /// This function is called in the start to add all the listeners to the UnityEvent
+    /// </summary>
+    public void AddListenersToEvent()
+    {
+        ResearchDoneEvent.AddListener(ResearchDone);
+        ResearchDoneEvent.AddListener(SetNextStatesInTree);
+        ResearchDoneEvent.AddListener(_nodeSetting.UnlockObjects);
+    }
+
+    /// <summary>
+    /// This function is called to update the UI for the node
+    /// </summary>
+    public void SetUIForStates()
+    {
+        switch (CurrentSkillState)
+        {
+            case SkillState.notAvailable:
+                _button.interactable = false;
+                gameObject.SetActive(false);
+                break;
+            case SkillState.available:
+                _button.interactable = true;
+                gameObject.SetActive(true);
+                break;
+            case SkillState.inDevelopment:
+                _button.interactable = false;
+                break;
+            case SkillState.bought:
+                _button.interactable = false;
+                break;
+        }
+    }
+
+    /// <summary>
+    /// This function is called to set the next nodes active when the research is done
+    /// </summary>
+    private void SetNextStatesInTree()
```

Comment:
Something tells me that the node should not care about the nodes it's connected to... If you think about a tree where the unlocking of a node depends on two or more nodes, this approach doesn't add up. Looking at the time and scope, you can leave it for now, but it's something you can think about